proposals/299-ip-failure-count.txt

```diff
@@ -156,20 +156,24 @@ Ticket: https://trac.torproject.org/projects/tor/ticket/27491
    One way we could forget old sessions is by halving all the failure point
    (FP) values before adding when:
 
-      * One or more failure point values are a multiple of a random number
-        between 1 and 5
+      * The failure point value being added to is a multiple of a random
+        number between 1 and 5
 
       * One or more failure point values are greater than or equal to 100
 
+   Halving by a random number as described above happens by first generating a
+   random number. Then if the FP value we are going to add to is a multiple of
+   the random number we generated, we halve.
+
    The reason for halving the values at regular intervals is to forget old
    sessions while keeping an approxmate history. We halve all FP values so
    that one IP version doesn't dominante on the failure count if the other
    is halved. This keeps an approximate scale of the failures on a client.
 
    The reason for halving at a multiple of a random number instead of a fixed
    interval is so we can halve regularly while not making it too predictable.
-   This prevents a situation where we would be halving too often to keep an
-   approximate failure history.
+   This prevents a situation where we would be halving too often where we can't
+   keep an approximate failure history.
```

Comment:
I don't understand this sentence.

When we halve too often, the numbers are zero.

When we don't halve often enough, the numbers are too large, and they can:
* block users from using tor after they change networks, or
* reveal user activity in the state file.

Which issue are you talking about here?

Reply:
I was talking about a (possibly hypothetical) situation where a client is unaware of which networks work as the failure values have been reduced to zero and then are reset.

Reply:
However, I can understand this will have issues with mobile clients.

Reply:
Clients should still be able to connect, even if they reset every time. History should be useful for optimisation, but not required.

Reply:
Makes sense.